cli/azd/design-notes/environments.md

```diff
@@ -0,0 +1,271 @@
+
+I think that with a few small tweaks to our model we can make it much easier to understand and describe - while at the same time allowing us to defer introduction of the concept until a user actually cares about the value an environment provides.
+
+## Proposed Model
+
+### Overview
+
+As before, we still have the concept of an environment in our system.  All `azd` commands still accept an `-e` parameter to select what environment is used, and `azd env select` can be used to select the default environment. However, if no environment is selected, `azd` creates a new environment with the name `default`. In addition to a name every environment gets a unique id (a v4 UUID) assigned to it. These values are recorded in `.azure/<environment-name>/config.json`:
+
+```json
+{
+  "id": "9fdfa455-2131-4faf-b2e3-8380aba97c96",
+  "name": "default"
+}
+```
+
+The above represents a fully initialized environment. Unlike today, creating an environment does not require user interaction and can be created without being logged into Azure.
+
+As before, an environment name needs to be unique with a project context (i.e. the `.azure` folder) ignoring case differences (useful when persisting data to file systems which may not allow case differences).
+
+### Deployment Parameters
+
+Bicep Templates can contain parameters, which are configured per deployment. Other IaC providers have similar models (config for Pulumi, variables for Terraform). Since values for these parameters are provided on every deployment there is often a way to specify them from a file, for example for Bicep, a *deployment parameters* file (also known as a `.parameters.json` file) may be used to configure these values.
+
+When `azd provision` is runs it loads the project wide configuration from the corresponding `.parameters.json` file in the infrastructure folder and merges with environment specific deployment parameters. If any required deployment parameters are missing the user is prompted for values and the responses are stored in the environment for future deployments.
+
+The deployment parameters are stored in `config.json` under the `infra` section, segmented by module.
+
+```json
+{
+  "id": "9fdfa455-2131-4faf-b2e3-8380aba97c96",
+  "name": "default",
+  "infra": {
+    "main": { 
+      "properties": {
+         // ... the object that is set as the value of the "properties" key of a deploymentParameters file
+      }
+    }
+  }
+}
+```
+
+To save ourselves the trouble of designing our own format, we reuse the schema of a `deploymentParamters` file. One nice feature here is that system provides enough flexibility to express references to secrets in Azure Key Vault, which will be a useful feature for us down the line.
+
+We create a new set of `azd infra config` commands with actions like `set`, `get`, `list`, `delete` to allow management of these parameters without having to edit the `config.json` file.
```

Comment:
In my opinion this area should be left up to the provision providers to implement how parameters are passed into the the respective provisioner.  In bicep things can stay the same - bicep will interpolate your template and replace with parameter values.

In terraform for example it can pass the parameters into the terraform plan using `--var` flags, etc.